bindings/matrix-sdk-ffi/src/widget.rs

```diff
@@ -0,0 +1,154 @@
+    pub read: Vec<EventFilter>,
+    /// Types of the messages that a widget wants to be able to send.
+    pub send: Vec<EventFilter>,
+}
+
+impl From<Permissions> for matrix_sdk::widget::Permissions {
+    fn from(value: Permissions) -> Self {
+        Self {
+            read: value.read.into_iter().map(Into::into).collect(),
+            send: value.send.into_iter().map(Into::into).collect(),
+        }
+    }
+}
+
+impl From<matrix_sdk::widget::Permissions> for Permissions {
+    fn from(value: matrix_sdk::widget::Permissions) -> Self {
+        Self {
+            read: value.read.into_iter().map(Into::into).collect(),
+            send: value.send.into_iter().map(Into::into).collect(),
+        }
+    }
+}
+
+/// Different kinds of filters that could be applied to the timeline events.
+#[derive(uniffi::Enum)]
+pub enum EventFilter {
+    /// Message-like events.
+    MessageLike {
+        /// The type of the message-like event.
+        event_type: String,
+        /// Additional filter for the msgtype, currently only used for
+        /// `m.room.message`.
+        msgtype: Option<String>,
+    },
+    /// State events.
+    State {
+        /// The type of the state event.
+        event_type: String,
+        /// State key that could be `None`, `None` means "any state key".
+        state_key: Option<String>,
+    },
+}
+
+impl From<EventFilter> for matrix_sdk::widget::EventFilter {
+    fn from(value: EventFilter) -> Self {
+        match value {
+            EventFilter::MessageLike { event_type, msgtype } => {
+                Self::MessageLike { event_type: event_type.into(), msgtype }
+            }
+            EventFilter::State { event_type, state_key } => {
+                Self::State { event_type: event_type.into(), state_key }
+            }
+        }
+    }
+}
+
+impl From<matrix_sdk::widget::EventFilter> for EventFilter {
+    fn from(value: matrix_sdk::widget::EventFilter) -> Self {
+        use matrix_sdk::widget::EventFilter as F;
+
+        match value {
+            F::MessageLike { event_type, msgtype } => {
+                Self::MessageLike { event_type: event_type.to_string(), msgtype }
+            }
+            F::State { event_type, state_key } => {
+                Self::State { event_type: event_type.to_string(), state_key }
+            }
+        }
+    }
+}
+
+#[uniffi::export(callback_interface)]
+pub trait PermissionsProvider: Send + Sync {
+    fn acquire_permissions(&self, permissions: Permissions) -> Permissions;
+}
+
+struct PermissionsProviderWrap(Arc<dyn PermissionsProvider>);
+
+#[async_trait]
+impl matrix_sdk::widget::PermissionsProvider for PermissionsProviderWrap {
+    async fn acquire_permissions(
+        &self,
+        permissions: matrix_sdk::widget::Permissions,
+    ) -> matrix_sdk::widget::Permissions {
+        let this = self.0.clone();
+        tokio::task::spawn_blocking(move || this.acquire_permissions(permissions.into()).into())
```

Comment:
I believe I understand why `spawn_blocking` is necessary here, but I would prefer to see a comment explaining it for our future-selves.

Is it because acquiring permissions can take time (the user has to _give_ permissions), so we don't want to block the entire thread while waiting on the user, right?